SETUP/upgrade/14/20200404_normalize_project_details.php

```diff
@@ -0,0 +1,69 @@
+<?php
+$relPath='../../../pinc/';
+include_once($relPath.'base.inc');
+include_once($relPath.'misc.inc');
+include_once($relPath.'unicode.inc');
+
+header('Content-type: text/plain');
+
+// ------------------------------------------------------------
+
+echo "Normalizing project details...";
+
+$sql = "
+    SELECT projectid, nameofwork, authorsname, comments, extra_credits
+    FROM projects
+    ORDER BY projectid;
+";
+
+$result = mysqli_query(DPDatabase::get_connection(), $sql) or die( mysqli_error(DPDatabase::get_connection()) );
+
+$projects = array();
+while($row = mysqli_fetch_assoc($result))
+{
+    $projects[] = $row;
+}
+
+$stop_file = '/tmp/utf8_normalization';
+
+echo <<<EOF
+This will now start normalizing all of the project fields.
+This will likely take a while.
+To stop the process gracefully create $stop_file
+(eg: touch $stop_file) and it will stop after it
+finishes the table it is working on.
+
+EOF;
+
+echo count($projects) . " possible projects to normalize.\n";
+
+foreach($projects as $project)
+{
+    if(file_exists($stop_file))
+    {
+        echo "\n\n";
+        echo "Detected stop file ($stop_file), so stopping before starting " . $project["projectid"] . ".\n";
+        echo "Delete this file and restart to continue.\n";
+        exit();
+    }
+
+    
+    $sql = "
+        UPDATE projects SET
+        nameofwork    = '".mysqli_real_escape_string(DPDatabase::get_connection(), utf8_normalize($project["nameofwork"]))."',
```

Comment:
For nameofwork, authorsname, and extra_credits it would be a big help to run these strings through `html_entity_decode()` first, so that any HTML entities are converted into their Unicode equiv first. Turns out we have several projects on PROD that have HTML entities in them. We *don't* want to do that for comments since it's expected those actually do have HTML entities we want to retain.

Reply:
There are a small handful of projects on DPUT that have entities in the titles, but none in the `authorsname` or `extra_credits` fields. Should I mung a few of them so that we can test all fields (probably not strictly necessary).

I agree that if PMs use entities in the Project Comments, they should be left as they are. What about the `postcomments`? I show none on TEST, but 38 in the `projects` table on PROD. By comparison, only 2 in the `extra_credits` field.

Reply:
As long as there are HTML entities in some fields in TEST that the script can convert for validation I don't think we need something in all of them -- so I wouldn't bother.

Ah, good questions. We should be normalizing `postcomments` as well, but I think we should treat them like we do `comments` as far as HTML entities (that is: leave them as entities).

Reply:
As long as they are treated like `comments` in terms of display, that should be fine. And yes, definitely normalize them, too!

Reply:
Ok updated on save and added postcomments too. However, I'm curious about the decoding stuff. Isn't that for html display? Why do we care if they have html like data in the database. Shouldn't that only be a display issue?

Reply:
It's okay if it's in the database for the `postcomments` and `comments` fields, but not for `nameofwork` and `authorsname`. I'm not sure about `extra_credits`, but I'd err on the side of No, unless we really want active links in that field. Do a project search on the test site for `<b>` or `<i>` in the title field. I think there's also `<b>` in the author field. See the results of what happens on the project page for this project: [Death's blue screen {test}](https://www.pgdp.org/c/project.php?id=projectID44dcfc8b0b855).

Reply:
(It's not the presence of the markup that messes with the display of the project page, but the fact that the element is not closed.)

Reply:
but should we just encode before rendering so that isn't an issue? that is how usually apps handle persisted XSS issues.

Reply:
Hm. Okay, I think I misunderstood your original question. Are you not questioning the fact that it needs to be done, but the order in which it should be done?

Reply:
That will be resolved with the changes that @cpeel backed out recently where we encode properly before displaying the name in the UI.

Reply:
Then I obviously went off on a complete tangent and don't know what I'm talking about!